### PR TITLE
Bed regions liftover cn fix

### PR DESCRIPTION
**Pull request by yunfeiguo**
- [x] [RES-453](https://binatechnologies.atlassian.net/browse/RES-453)
- [x] Blocking PRs: ADD HERE IF ANY
* People
  - [x] Reviewers (>=1): @marghoob 
  - [x] Merger (>=1): @marghoob 
  - [x] FYI: @ulitskyb
- [x] Release note (if needed)
- [x] No code copied from outside Bina 
- [x] Tests (Unit, Workflow, ITL & Gherkin) are passing
- [ ] PR labels, milestones & assignees

# Summary

* added a simple function to convert CN from String to single integer back and forth

### CODE REVIEW

**Comment by marghoob** [comment]
@yunfeiguo can you merge base:master into your branch here so that the diff shows up better?

**Comment by yunfeiguo** [comment]
@marghoob done. Thanks!

File: varsim.py [hunk 1]
```diff
@@ -21,6 +24,51 @@
...
+    convert '2/1'-like copy number to a single number(e.g. 2)
+    0 will be considered same as 1
+    by default the max number will be kept
+    the change is in place
+    """
+    if operation != "two2one" and operation != "one2two":
+        raise ValueError("Only two2one or one2two allowed")
+    two2one = operation == "two2one"
+    for name in filenames:
+        file = open(name, "r")
```
Comment (marghoob) [inline varsim.py]: Please do not use `file` as it is a known type in Python. You can try something like `with open(name, "r") as file_fd:` . Using `with` block will also nicely close the file when the block is exited.

File: varsim.py [hunk 2]
```diff
@@ -21,6 +24,51 @@
...
+        output = tempfile.NamedTemporaryFile(mode = 'r+w', delete = False)
+        for l in file:
+            l = string.rstrip(l)
+            fields = string.split(l, "\t")
```
Comment (marghoob) [inline varsim.py]: You can use `fields = l.rstrip().split("\t")` since split is a member function of the string class.

File: varsim.py [hunk 3]
```diff
@@ -21,6 +24,51 @@
...
+            if string.find(l,"#") == 0 or string.find(fields[8],"CN") < 0:
```
Comment (marghoob) [inline varsim.py]: Is this checking for a comment line? Use `fields[0].startswith("#")`. For copy number, you can use `"CN" not in fields[8]` which will check for substring containment.

File: varsim.py [hunk 4]
```diff
@@ -21,6 +24,51 @@
...
+                output.write(l + "\n")
+            else:
+                info = string.split(fields[8], ':')
```
Comment (marghoob) [inline varsim.py]: `fields[8].split(":")`

File: varsim.py [hunk 5]
```diff
@@ -21,6 +24,51 @@
...
+                cnIndex = info.index('CN')
+                gtIndex = info.index('GT')
+                #change CN field in all samples
+                for sampleIndex in range(9,len(fields)):
+                    sampleInfo = string.split(fields[sampleIndex], ':')
+                    if two2one:
+                        cn = re.compile('[/|]').split(sampleInfo[cnIndex])
```
Comment (marghoob) [inline varsim.py]: The pattern `re.compile("[/|]")` can be a constant compiled once instead of in every loop iteration.

Comment (yunfeiguo) [inline varsim.py]: All done. Thanks!

File: varsim.py [hunk 6]
```diff
@@ -21,6 +23,51 @@
...
+                            cn = sampleInfo[cnIndex]
+                            for i in range(len(gt)):
+                                gt[i] = '1' if gt[i] == '0' else cn
+                            if sampleInfo[gtIndex].find('/') >= 0:
+                                sampleInfo[cnIndex] = '/'.join(gt)
+                            else:
+                                sampleInfo[cnIndex] = '|'.join(gt)
+                        fields[sampleIndex] = ":".join(sampleInfo)
+                    output.write("\t".join(fields) + "\n")
+        output.close()
```
Comment (marghoob) [inline varsim.py]: `output` is being closed outside of the block where it was initialized.

File: varsim.py [hunk 7]
```diff
@@ -21,6 +23,51 @@
...
+    if operation != "two2one" and operation != "one2two":
+        raise ValueError("Only two2one or one2two allowed")
+    two2one = operation == "two2one"
+    delimiter = re.compile('[/|]')
+    for name in filenames:
+	with open(name, 'r') as file_fd:
+            output = tempfile.NamedTemporaryFile(mode = 'r+w', delete = False)
+            for l in file_fd:
+		l = l.rstrip()
+                fields = l.split("\t")
```
Comment (marghoob) [inline varsim.py]: Looks like the indentation is an extra space when compared to the previous line.

File: varsim.py [hunk 8]
```diff
@@ -21,6 +23,51 @@
 REQUIRE_VARSIMJAR = not os.path.isfile(DEFAULT_VARSIMJAR)
 if REQUIRE_VARSIMJAR: DEFAULT_VARSIMJAR = None
 
+def convertCN(filenames, operation):
```
Comment (marghoob) [inline varsim.py]: This function does not seem to fix the VCF header accordingly to match the interpretation of CN

File: varsim.py [hunk 9]
```diff
@@ -21,6 +23,56 @@
...
+                else:
+                    info = fields[8].split(':')
+                    cnIndex = info.index('CN')
+                    gtIndex = info.index('GT')
+                    #change CN field in all samples
+                    for sampleIndex in range(9,len(fields)):
+                        sampleInfo = fields[sampleIndex].split(':')
+                        if two2one:
+                            cn = delimiter.split(sampleInfo[cnIndex])
+                            sampleInfo[cnIndex] = str(max(cn))
```
Comment (marghoob) [inline varsim.py]: Taking `max` over a string array does not seem right. Don't you need to convert `cn` to a list of ints?

Comment (yunfeiguo) [inline varsim.py]: Oops~ Sorry. Fixed.

**Comment by yunfeiguo** [comment]
Thanks! Fixed.

**Comment by marghoob** [comment]
Thanks for the patience with comments. Merging now. 👍

**Comment by yunfeiguo** [comment]
Haha, I should thank you for pointing out potential improvements! I really need to learn more about python programming.

**Comment by yunfeiguo** [comment]
Happy Thanksgiving!

**Comment by marghoob** [comment]
You too have a great thanksgiving! Give yourself a good rest and enjoy the
vacation!

**Pull request merged**